src/main/cpp/src/zorder.cu

```diff
@@ -27,6 +27,118 @@
 #include <thrust/iterator/constant_iterator.h>
 #include <thrust/iterator/counting_iterator.h>
 
+namespace {
+
+// pretends to be an array of uint32_t, but really only stores
+// the data in a long with a set number of bits allocated for
+// each item
+struct long_backed_array {
+  long_backed_array() = delete;
+  ~long_backed_array() = default;
+  long_backed_array(long_backed_array const&) = default;  ///< Copy constructor
+  long_backed_array(long_backed_array&&) = default;  ///< Move constructor
+  inline __device__ explicit long_backed_array(int32_t num_bits): data(0), 
+    num_bits(num_bits),  mask(static_cast<uint64_t>((1L << num_bits) - 1)) {}
+
+  /**
+   * @brief Copy assignment operator
+   *
+   * @return Reference to this object
+   */
+  long_backed_array& operator=(long_backed_array const&) = default;
+  /**
+   * @brief Move assignment operator
+   *
+   * @return Reference to this object (after transferring ownership)
+   */
+  long_backed_array& operator=(long_backed_array&&) = default;
+
+  inline __device__ uint32_t operator[](int i) const {
+    int32_t offset = num_bits * i;
+    return (data >> offset) & mask;
+  }
+
+  inline __device__ void set(int i, uint32_t value) {
+    int32_t offset = i * num_bits;
+    uint64_t masked_data = data & ~(static_cast<uint64_t>(mask) << offset);
+    data = masked_data | (static_cast<uint64_t>(value & mask) << offset);
+  }
+
+private:
+  uint64_t data;
+  int32_t num_bits;
+  uint32_t mask;
+};
+
+
+// Most of the hilbert index code is based off of the work done by David Moten at
+// https://github.com/davidmoten/hilbert-curve, which has the following Note in
+// the code too
+// This algorithm is derived from work done by John Skilling and published
+// in "Programming the Hilbert curve". (c) 2004 American Institute of Physics.
+// With thanks also to Paul Chernoch who published a C# algorithm for Skilling's
+// work on StackOverflow and
+// <a href="https://github.com/paulchernoch/HilbertTransformation">GitHub</a>.
+__device__ uint64_t to_hilbert_index(const long_backed_array & transposed_index, const int num_bits, const int num_dimensions) {
```

Comment:
It is the same `num_bits`. I called it this to match closely with the original code.

https://github.com/davidmoten/hilbert-curve/blob/758058c0b86ff01edffb46317f04a619d861ba14/src/main/java/org/davidmoten/hilbert/HilbertCurve.java#L26

but I added `num_` to it because I thought `bits` was not descriptive enough. Happy to make it clearer.